sky/packages/sky/lib/src/fn3/drag_target.dart

```diff
@@ -3,15 +3,85 @@
 // found in the LICENSE file.
 
 import 'dart:collection';
+import 'dart:sky' as sky;
 
 import 'package:sky/rendering.dart';
 import 'package:sky/src/fn3/basic.dart';
 import 'package:sky/src/fn3/binding.dart';
 import 'package:sky/src/fn3/framework.dart';
+import 'package:sky/src/fn3/navigator.dart';
 
 typedef bool DragTargetWillAccept<T>(T data);
 typedef void DragTargetAccept<T>(T data);
 typedef Widget DragTargetBuilder<T>(BuildContext context, List<T> candidateData, List<dynamic> rejectedData);
+typedef void DragFinishedNotification();
+
+class Draggable extends StatefulComponent {
+  Draggable({ Key key, this.navigator, this.data, this.child, this.feedback }): super(key: key) {
+    assert(navigator != null);
+  }
+  final NavigatorState navigator;
+  final dynamic data;
+  final Widget child;
+  final Widget feedback;
+  DraggableState createState() => new DraggableState();
+}
+
+class DraggableState extends State<Draggable> {
+  DragRoute _route;
+
+  void _startDrag(sky.PointerEvent event) {
+    if (_route != null)
+      return; // TODO(ianh): once we switch to using gestures, just hand the gesture to the route so it can do everything itself. then we can have multiple drags at the same time.
+    Point point = new Point(event.x, event.y);
+    RenderBox renderObject = context.findRenderObject();
+    _route = new DragRoute(
+      data: config.data,
+      dragStartPoint: renderObject.globalToLocal(point),
+      feedback: config.feedback,
+      onDragFinished: () {
+        _route = null;
+      }
+    );
+    _route.update(point);
+    config.navigator.push(_route);
+  }
+
+  void _updateDrag(sky.PointerEvent event) {
+    if (_route != null) {
+      config.navigator.setState(() {
```

Comment:
Yeah, totally. I'm not really sure what to do about it. It's not the only place we do this.